runtime/lmsig_ksi-ls12.c

```diff
@@ -0,0 +1,329 @@
+				rsksiSetBlockLevelLimit(pThis->ctx, pvals[i].val.d.n);
+			}
+			bHaveLevelLimit = true;
+
+		} else if (!strcmp(pblk.descr[i].name, "sig.block.timeLimit")) {
+			if (pvals[i].val.d.n < 0) {
+				errmsg.LogError(0, RS_RET_ERR, "sig.block.timeLimit "
+					"%llu invalid - signing disabled", pvals[i].val.d.n);
+				pThis->ctx->disabled = true;
+			} else {
+				rsksiSetBlockTimeLimit(pThis->ctx, pvals[i].val.d.n);
+			}
+		} else if (!strcmp(pblk.descr[i].name, "sig.keeprecordhashes")) {
+			rsksiSetKeepRecordHashes(pThis->ctx, pvals[i].val.d.n);
+		} else if(!strcmp(pblk.descr[i].name, "sig.keeptreehashes")) {
+			rsksiSetKeepTreeHashes(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "sig.syncmode")) {
+			cstr = (uchar*) es_str2cstr(pvals[i].val.d.estr, NULL);
+			if (!strcasecmp((char*) cstr, "sync")) rsksiSetSyncMode(pThis->ctx, LOGSIG_SYNCHRONOUS);
+			else if (!strcasecmp((char*) cstr, "async")) rsksiSetSyncMode(pThis->ctx, LOGSIG_ASYNCHRONOUS);
+			else errmsg.LogError(0, RS_RET_ERR, "sig.syncmode '%s' unknown - using default", cstr);
+			free(cstr);
+		} else if (!strcmp(pblk.descr[i].name, "sig.randomsource")) {
+			cstr = (uchar*) es_str2cstr(pvals[i].val.d.estr, NULL);
+			rsksiSetRandomSource(pThis->ctx, (char*) cstr);
+			free(cstr);
+		} else if (!strcmp(pblk.descr[i].name, "dirowner")) {
+			rsksiSetDirUID(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "dirownernum")) {
+			rsksiSetDirUID(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "dirgroup")) {
+			rsksiSetDirGID(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "dirgroupnum")) {
+			rsksiSetDirGID(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "fileowner")) {
+			rsksiSetFileUID(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "fileownernum")) {
+			rsksiSetFileUID(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "filegroup")) {
+			rsksiSetFileGID(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "filegroupnum")) {
+			rsksiSetFileGID(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "dircreatemode")) {
+			rsksiSetDirCreateMode(pThis->ctx, pvals[i].val.d.n);
+		} else if (!strcmp(pblk.descr[i].name, "filecreatemode")) {
+			rsksiSetCreateMode(pThis->ctx, pvals[i].val.d.n);
+		} else {
+			DBGPRINTF("lmsig_ksi: program error, non-handled "
+			  "param '%s'\n", pblk.descr[i].name);
+		}
+	}
+
+	if (!ag_uri)
+		REPORT_PARAM_MISSING("sig.aggregator.url");
+
+	if (!ag_loginid)
+		REPORT_PARAM_MISSING("sig.aggregator.user");
+
+	if (!ag_key)
+		REPORT_PARAM_MISSING("sig.aggregator.key");
+
+	if (!bHaveLevelLimit)
+		REPORT_PARAM_MISSING("sig.aggregator.levelLimit");
+
+	if (!bHaveHashFunction)
+		rsksiSetHashFunction(pThis->ctx, "SHA2-256");
+
+	if (rsksiSetAggregator(pThis->ctx, ag_uri, ag_loginid, ag_key) != KSI_OK) {
+		pThis->ctx->disabled = true;
```

Comment:
The error message is emitted inside rsksiSetAggregator where proper error code is available